servers/src/main/java/tachyon/master/next/IndexedSet.java

```diff
@@ -0,0 +1,247 @@
+   * o.equals(o2)). If this set already contains the object, the call leaves the set unchanged.
+   *
+   * @param object the object to add
+   */
+  public void add(T object) {
+    synchronized (mFields) {
+      for (String field : mIndexMap.keySet()) {
+        Map<Object, Set<T>> fieldValueToSet = mSetIndexedByFieldValue.get(mIndexMap.get(field));
+        Object value = getField(object, field);
+        if (fieldValueToSet.containsKey(value)) {
+          fieldValueToSet.get(value).add(object);
+        } else {
+          Set<T> set = new HashSet<T>();
+          set.add(object);
+          fieldValueToSet.put(value, set);
+        }
+      }
+    }
+  }
+
+  /**
+   * @return a set of all objects
+   */
+  public Set<T> all() {
+    Map<Object, Set<T>> setForOneField;
+    synchronized (mFields) {
+      setForOneField = mSetIndexedByFieldValue.get(0);
+    }
+    Set<T> ret = new HashSet<T>();
+    for (Set<T> set : setForOneField.values()) {
+      ret.addAll(set);
+    }
+    return ret;
+  }
+
+  /**
+   * Whether there is an object with the specified field value in the set.
+   *
+   * @param index the field index
+   * @param value the field value
+   * @return true if there is one such object, otherwise false
+   */
+  public boolean contains(FieldIndex index, Object value) {
+    return !getByField(index, value).isEmpty();
+  }
+
+  /**
+   * Get a subset of objects with the specified field value.
+   *
+   * @param index the field index
+   * @param value the field value to be satisfied
+   * @return the set of objects or an empty set if no such object exists
+   */
+  public Set<T> getByField(FieldIndex index, Object value) {
+    int id = getIndex(index);
+    Set<T> set;
+    synchronized (mFields) {
+      set = mSetIndexedByFieldValue.get(id).get(value);
+    }
+    return set == null ? new HashSet<T>() : set;
+  }
+
+  /**
+   * Get the first object from the set of objects with the specified field value.
+   *
+   * @param index the field index
+   * @param value the field value
+   * @return the object or null if there is no such object
+   */
+  public T getFirst(FieldIndex index, Object value) {
+    Set<T> all = getByField(index, value);
+    return all.isEmpty() ? null : all.iterator().next();
+  }
+
+  /**
+   * Remove an object from the set.
+   *
+   * @param object the object to remove
+   * @return true if removed successfully, otherwise false
+   */
+  public boolean remove(T object) {
+    boolean success = true;
+    for (String field : mIndexMap.keySet()) {
+      Object fieldValue = getField(object, field);
+      int id = mIndexMap.get(field);
+      synchronized (mFields) {
+        Set<T> set = mSetIndexedByFieldValue.get(id).remove(fieldValue);
+        if (set != null) {
+          success = success && set.remove(object);
+          if (!set.isEmpty()) {
+            mSetIndexedByFieldValue.get(id).put(fieldValue, set);
```

Comment:
This pattern is a bit unusual to me:
you remove the `set` of `fieldValue` first, remove `object` from `set`, and put `set` back to `mSetIndexedByFieldValue`.
So it gets two mutation for each `remove(T object)`. Can we simply get the `set` by `fieldValue`, remove `object`, and remove this `set` only if `set` becomes empty? It seems a bit more efficient compared to the current code.